File: services/server/src/protocol/message.py
TYPE_BETS = 1
TYPE_END = 2

BATCH_HEADER_LEN = 2
BET_HEADER_LEN = 2
MIN_BET_LEN = 20


def parse_message_type(packet: bytes) -> int:
    message_type = packet[0]
    if message_type not in (TYPE_BETS, TYPE_END):
        raise ValueError(f"invalid message type: {message_type}")

    return message_type


def serialize_end() -> bytes:
    payload = bytearray([TYPE_END])
    return len(payload).to_bytes(2, byteorder="big") + bytes(payload)


def is_end_packet(packet: bytes) -> bool:
    return len(packet) == 2 and packet[0] == TYPE_END

File: services/client/src/protocol/message.go
package protocol

import (
	"encoding/binary"
	"fmt"
)

const (
	TYPE_BETS byte = 1
	TYPE_END  byte = 2
	TYPE_ACK  byte = 3

	BATCH_HEADER_LEN = 2
	BET_HEADER_LEN   = 2
	MIN_BET_LEN      = 20
)

func is_valid_type(t byte) bool {
	return t == TYPE_BETS || t == TYPE_END || t == TYPE_ACK
}

func ParseMessageType(packet []byte) (byte, error) {
	messageType := packet[0]
	if !is_valid_type(messageType) {
		return 0, fmt.Errorf("invalid message type: %d", packet[0])
	}

	return messageType, nil
}

func SerializeEnd(agencyID uint8) []byte {
	packet := make([]byte, 0, 4)
	packet = binary.BigEndian.AppendUint16(packet, 2)
	packet = append(packet, TYPE_END, agencyID)
	return packet
}

func SerializeAck(agencyID uint8) []byte {
	packet := make([]byte, 0, 4)
	packet = binary.BigEndian.AppendUint16(packet, 2)
	packet = append(packet, TYPE_ACK, agencyID)
	return packet
}

func IsEndPacket(data []byte) bool {
	return len(data) == 1 && data[0] == TYPE_END
}

func IsAckPacket(data []byte) bool {
	return len(data) == 2 && data[0] == TYPE_ACK
}
